Upgrade Elasticsearch client and drop support for unsupported versions

hashhar opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

Elasticsearch clients are forward compatible and only guaranteed to work
with servers with matching major versions. i.e. 6.8.23 client would work
with 6.8.x and newer servers only. This also changes the
TestElasticsearch6ConnectorTest to use 6.6.0 version for tests since
that's the oldest version against which the connector passes all tests.
    
This also adds comments explaining why an exclusion exists and also
removes those that are no longer required - reducing the chances that
someone runs into runtime errors if they do something which we don't
explicitly test.

## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# Elasticsearch
* Drop support for Elasticsearch servers older than 6.6.0. ({issue}`11263`)
```

hashhar commented:
I've removed all exclusions for now to see what fails and will clean up exclusions as required.

hashhar commented:
cc: @martint Somehow it's working without any exclusions applied. Can you share some context on why the exclusions were added and if they may still be needed?

Also I saw that even 6.8 got EOLed last month, so maybe in a few months we'll want to drop 6.8.x support for good as well. I've kept it here for now.

martint commented:
> Can you share some context on why the exclusions were added and if they may still be needed?

There was a conflict with dependencies brought in by other components. It's possible those are already updated and the conflict no longer exists.

hashhar commented:
I'll check the dependency tree before and after the change. Thanks for sharing context @martint.

Would we want to drop 6.x altogether since it went EOL Feb 2022 or keep 6.8 for some time and then drop sometime in future?

martint commented:
> Would we want to drop 6.x altogether since it went EOL Feb 2022 or keep 6.8 for some time and then drop sometime in future?

We should poll users to see what version they are running before dropping support for any version.

phd3 commented:
> This also changes the
TestElasticsearch6ConnectorTest to use 6.6.0 version for tests since
that’s the oldest version against which the connector passes all tests.

which tests are we talking about here? is this for a new pr?

hashhar commented:
The "tests" there refers to all tests in the `trino-elasticsearch` module. The client version is being bumped because of some CVEs and because 6.x has been unsupported for some time now - 6.8.23 is the last 6.x version.

phd3 reviewed (approved):
looks good % https://github.com/trinodb/trino/pull/11263#issuecomment-1058372017 taken care of

hashhar commented on plugin/trino-elasticsearch/pom.xml:
```diff
@@ -433,13 +379,24 @@
                             <requireUpperBoundDeps>
                                 <excludes combine.children="append">
-                                    <!-- trino-main depends on a newer version -->
+                                    <!-- trino-main depends on 8.4.1 while org.elasticsearch:elasticsearch:6.8.23 depends on 7.7.3 -->
```
I'll send a separate PR to clean up exclusions in other modules as well - I suspect we have lot of un-needed exclusions over time.

hashhar commented on plugin/trino-elasticsearch/pom.xml:
```diff
@@ -162,7 +145,7 @@
         <dependency>
             <groupId>org.apache.httpcomponents</groupId>
             <artifactId>httpcore</artifactId>
-            <version>4.4.5</version>
```
Bumping httpcore-nio also seems to work. I originally bumped only httpcore since the enforcer plugin complained about only that.

hashhar commented:
> We should poll users to see what version they are running before dropping support for any version.

See https://trinodb.slack.com/archives/CFLB9AMBN/p1655477455137409, https://trinodb.slack.com/archives/CP1MUNEUX/p1655450446249759

@martint @phd3 Please take a look, addressed comments.

FYI @bitsondatadev @mosabua

phd3 commented on plugin/trino-elasticsearch/pom.xml:
```diff
@@ -180,74 +163,37 @@
             </exclusions>
         </dependency>
 
+        <dependency>
```
can you please remind why these need to be added?

hashhar replied:
Also example of bug fixed by newer client - https://github.com/elastic/elasticsearch/pull/29230 + https://github.com/elastic/elasticsearch/issues/40511

phd3 reviewed (approved):
Thanks! may be add a one-liner in commit description on why we're bumping client version.

hashhar commented:
Amended just the commit message - PTAL @phd3.

mosabua commented:
We should definitely add this in the release notes and @colebow and myself can also mention it in the Trino Community Broadcast
